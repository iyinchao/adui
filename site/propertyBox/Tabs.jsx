import React from "react"
import styles from "./components/propertyBox.scss"
import CodeBox from "./components/CodeBox"
import { Button, Switch, Tabs } from "../../components"
import { SIZES } from "./consts"

export default class PropertyBox extends React.Component {
  state = {
    banner: false,
    size: "small",
  }

  handleUpdate = () => {
    /**
     * Tabs 组件为了防止性能消耗，只会在 value 变化时触发方法重新计算
     * 这里 Propertybox 有点特殊，因此以这样的方式强制调用
     * 并非组件 bug
     */
    setTimeout(() => {
      this.tabs.updateIndicatorStyle()
    }, 150)
  }

  render() {
    const { banner, size } = this.state

    const codes = `<Tabs 
  ${banner ? `banner` : ""}
  defaultValue={0} 
  ${size === "small" ? "" : `size="${size}"`}
>
  <Tabs.Tab title="选项一" value={0} />
  <Tabs.Tab title="选项二" value={1} />
  <Tabs.Tab title="选项三" value={2} />
  <Tabs.Tab title="禁用选项" value={3} disabled />
</Tabs>`

    return (
      <div className={styles.wrapper}>
        <div className={styles.property}>
          <div className={styles.left}>
            <Tabs
              defaultValue={0}
              size={size}
              banner={banner}
              style={banner ? { flex: 1 } : null}
              ref={(tabs) => {
                this.tabs = tabs
              }}
            >
              <Tabs.Tab title="选项一" value={0} />
              <Tabs.Tab title="选项二" value={1} />
              <Tabs.Tab title="选项三" value={2} />
              <Tabs.Tab title="禁用选项" value={3} disabled />
            </Tabs>
          </div>
          <div className={styles.right}>
            <div className={styles.rightTitle}>可配置项</div>
            <dl className={styles.controls}>
              <dt className={styles.controlTitle}>尺寸</dt>
              <dd className={styles.controlContent}>
                <Button.Group banner>
                  {SIZES.map(({ value, text }) => (
                    <Button
                      key={value}
                      active={value === size}
                      onClick={() =>
                        this.setState({ size: value }, this.handleUpdate)
                      }
                    >
                      {text}
                    </Button>
                  ))}
                </Button.Group>
              </dd>
            </dl>
            <div className={styles.switchControl} style={{ marginTop: "20px" }}>
              <Switch
                checkedText="填满宽度 (banner)"
                unCheckedText="填满宽度 (banner)"
                checked={banner}
                onChange={(param) =>
                  this.setState({ banner: param }, this.handleUpdate)
                }
              />
            </div>
          </div>
        </div>
        <CodeBox codes={codes} />
      </div>
    )
  }
}
