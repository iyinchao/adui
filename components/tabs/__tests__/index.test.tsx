import * as React from "react"
import { act } from "react-dom/test-utils"
import TestRenderer from "react-test-renderer"
import { mount } from "enzyme"
import Tabs from ".."

window.ResizeObserver =
  window.ResizeObserver ||
  jest.fn().mockImplementation(() => ({
    disconnect: jest.fn(),
    observe: jest.fn(),
    unobserve: jest.fn(),
  }))

const { Tab } = Tabs

describe("Tabs", () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it("验证快照", () => {
    const wrapper0 = TestRenderer.create(
      <Tabs gutter={10}>
        <Tab title="1" value="1" />
      </Tabs>
    )
    const wrapper1 = TestRenderer.create(<Tabs>string</Tabs>)
    expect(wrapper0).toMatchSnapshot()
    expect(wrapper1).toMatchSnapshot()
  })

  it("验证 setIndicatorStyle 快照", () => {
    const wrapper0 = mount(
      <Tabs value="2">
        <Tab title="1" value="1" />
        <Tab title="2" value="2" />
      </Tabs>
    )
    const wrapper1 = mount(
      <Tabs>
        <Tab title="1" value="1" />
        <Tab title="2" value="2" />
      </Tabs>
    )
    act(() => {
      // 等待 0ms，这样才能触发 setIndicatorStyle 方法
      jest.advanceTimersByTime(0)
    })
    expect(wrapper0).toMatchSnapshot()
    expect(wrapper1).toMatchSnapshot()
  })

  it("是否能够内部驱动", () => {
    const wrapper = mount(
      <Tabs defaultValue="1">
        <Tab title="1" value="1" />
        <Tab title="2" value="2" />
      </Tabs>
    )
    expect(wrapper.find("div").at(0).props()["data-value"]).toEqual("1")
    wrapper.find("[role='tab']").at(1).simulate("click")
    expect(wrapper.find("div").at(0).props()["data-value"]).toEqual("2")
  })

  it("是否能受外部控制", () => {
    const wrapper0 = mount(
      <Tabs value="1">
        <Tab title="1" value="1" />
        <Tab title="2" value="2" />
      </Tabs>
    )
    expect(wrapper0.find("div").at(0).props()["data-value"]).toEqual("1")
    wrapper0.find("[role='tab']").at(1).simulate("click")
    expect(wrapper0.find("div").at(0).props()["data-value"]).toEqual("1")
  })
})
