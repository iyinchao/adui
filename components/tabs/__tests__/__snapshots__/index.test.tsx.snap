// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs 验证 setIndicatorStyle 快照 1`] = `ReactWrapper {}`;

exports[`Tabs 验证 setIndicatorStyle 快照 2`] = `ReactWrapper {}`;

exports[`Tabs 验证快照 1`] = `
<div
  className="adui-tabs-base adui-tabs-small"
>
  <div
    className="adui-tabs-indicator"
    style={Object {}}
  />
  <div
    aria-disabled={false}
    aria-selected={false}
    className="adui-tabs-tab"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="tab"
    style={
      Object {
        "marginRight": "10px",
      }
    }
    tabIndex={-1}
  >
    1
  </div>
</div>
`;

exports[`Tabs 验证快照 2`] = `
<div
  className="adui-tabs-base adui-tabs-small"
>
  <div
    className="adui-tabs-indicator"
    style={Object {}}
  />
  string
</div>
`;
